src/library/scala/collection/immutable/List.scala

```diff
@@ -556,6 +556,45 @@ sealed abstract class List[+A]
     result
   }
 
+  override def partition(p: A => Boolean): (List[A], List[A]) = {
+    if (isEmpty) List.TupleOfNil
+    else {
+      var b0 = true
+      var canConserve = true
+      var ys = this
+      var ayes: ListBuffer[A] = null
+      var nays: ListBuffer[A] = null
+      var n = 0
+      while (!ys.isEmpty) {
+        val y = ys.head
+        val b = p(y)
+        if (n == 0) b0 = b
+        val newCanConserve = canConserve && b == b0
+        if (canConserve && !newCanConserve) {
+          ayes = new ListBuffer[A]
+          nays = new ListBuffer[A]
+          val prefix = if (b0) ayes else nays
+          var j = 0
+          var zs = this
+          while (j < n) {
+            prefix += zs.head
+            zs = zs.tail
+            j += 1
+          }
+          canConserve = newCanConserve
+        }
+        if (!canConserve)
+          (if (b) ayes else nays) += y
+        n += 1
+        ys = ys.tail
```

Comment:
May I suggest the following rearrangement, where most of the loop body is gated by an initial `if (canConserve)`:
```scala
      while (!ys.isEmpty) {
        val y = ys.head
        val b = p(y)
        if (canConserve) {
          if (n == 0) b0 = b
          else if (b != 0) {
            canConserve = false
            ayes = new ListBuffer[A]
            nays = new ListBuffer[A]
            val prefix = if (b0) ayes else nays
            var j = 0
            var zs = this
            while (j < n) {
              prefix += zs.head
              zs = zs.tail
              j += 1
            }
            (if (b) ayes else nays) += y
          }
          n += 1
        } else {
          (if (b) ayes else nays) += y
        }
        ys = ys.tail
      }
```